lock/etcd.go

```
@@ -0,0 +1,134 @@
+	srv := &etcdSrv{
+		cfg: cfg,
+		m:   &sync.Mutex{},
+	}
+	return srv
+}
+
+type etcdSrv struct {
+	cfg    clientv3.Config
+	client *clientv3.Client
+	m      *sync.Mutex
+	closed bool
+}
+
+func (s *etcdSrv) connect() error {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	if s.client == nil {
+		client, err := clientv3.New(s.cfg)
+		if err != nil {
+			return err
+		}
+
+		s.client = client
+	}
+
+	return nil
+}
+
+func (s *etcdSrv) NewSession(cfg *SessionConfig) (Session, error) {
+	if err := s.connect(); err != nil {
+		return nil, err
+	}
+
+	ttl := int(math.Ceil(cfg.TTL.Seconds()))
+	session, err := concurrency.NewSession(s.client, concurrency.WithTTL(ttl))
+	if err != nil {
+		return nil, err
+	}
+
+	return &etcdSess{
+		cfg:     cfg,
+		session: session,
+	}, nil
+}
+
+func (l *etcdSrv) Close() error {
+	if l.closed {
+		return ErrAlreadyClosed.New()
+	}
+
+	defer func() { l.closed = true }()
+
+	if l.client == nil {
+		return nil
+	}
+
+	return l.client.Close()
+}
+
+type etcdSess struct {
+	cfg     *SessionConfig
+	session *concurrency.Session
+}
+
+func (s *etcdSess) NewLocker(id string) Locker {
+	return &etcdLock{
+		cfg:     s.cfg,
+		session: s.session,
+		mutex:   concurrency.NewMutex(s.session, id),
+	}
+}
+
+func (s *etcdSess) Close() error {
+	session := s.session
+	if session == nil {
+		return ErrAlreadyClosed.New()
+	}
+
+	s.session = nil
+	return session.Close()
+}
+
+type etcdLock struct {
+	cfg     *SessionConfig
+	session *concurrency.Session
+	mutex   *concurrency.Mutex
+}
+
+func (l *etcdLock) Lock() (<-chan struct{}, error) {
+	ctx := context.Background()
+
+	ttl := l.cfg.TTL
+	if ttl > 0 {
+		if ttl < time.Second {
+			ttl = time.Second
+		}
+
+		ctx, _ = context.WithTimeout(ctx, ttl)
+	}
+
+	if err := l.mutex.Lock(ctx); err != nil {
+		if isContextDeadlineExceededError(err) {
+			return nil, ErrCanceled.Wrap(err)
+		}
+
+		return nil, err
+	}
+
+	return ctx.Done(), nil
+}
+
+func (m *etcdLock) Unlock() {
+	_ = m.mutex.Unlock(context.TODO())
```

this does not appear to be right for me, we should return the error if any.